### PR TITLE
Workflow Mutex Sample

### PR DESCRIPTION
**Pull request by devbased**
## What was changed
Added a new sample showcasing how to implement a Mutex Workflow for managing concurrent access to a shared resource.

~_NOTE: Since the required SDK version (the one with [workflow history info](https://github.com/temporalio/sdk-dotnet/commit/9b927098be7e277d688798b0da235fd2182643ca)) hasn't been published yet, I had to add links to the SDK in TemporalioSamples.sln and Directory.build.props_~

### CODE REVIEW

File: Directory.Build.props
Comment (cretz) [inline Directory.Build.props:29]: :+1: Will leave this PR unmerged until the next SDK release

File: src/Mutex/Impl/ILockHandle.cs
Comment (cretz) [inline src/Mutex/Impl/ILockHandle.cs:1]: I think `ILockHandle` and `WorkflowMutex` can move up to the `TemporalioSamples.Mutex` with the rest (can make former inner class of latter if wanting fewer files). And can probably move `MutexWorkflow.workflow.cs` up there too and put most things it uses as inner classes if we want to avoid file sprawl.

Your call, but I think there's value in fewer files/namespaces here

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:8]: Not sure there is value for a separate lock handler concept instead of just embedding logic in this class which is the only place that uses it

File: src/Mutex/Impl/Internal/MutexActivities.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexActivities.cs:32]: I'm thinking I should make a more typesafe form of this as a helper on the options

File: src/Mutex/Impl/Internal/MutexActivities.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexActivities.cs:29]: I think the workflow ID should have something prepended just in case they are using this resource ID as a workflow ID elsewhere

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:12]: Don't have to do it, but there's a neat feature in .NET SDK not in other SDKs - you can have a constructor with `[WorkflowInit]` that accepts this input there

Comment (cretz) [inline src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:12]: Often clearer to accept the non-null input and make your caller just pass an empty form of it. You're the only caller of this, so you can pass an empty input.

File: src/Mutex/Impl/Internal/MutexWorkflowInput.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexWorkflowInput.cs:3]:
```suggestion
internal record MutexWorkflowInput(IReadOnlyCollection<LockRequest> InitialRequests);
```

I have been discouraging enumerable as parameters in this use case because of the confusion with users when they expect it to be lazy or not re-enumerable.

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [inline src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:32]: So it looks like this can only handle one lock at a time which I am not sure is what we want

File: src/Mutex/README.md
Comment (cretz) [inline src/Mutex/README.md:1]: Don't forget to update main README with reference to this one

File: src/Mutex/WorkflowWithMutex.workflow.cs
Comment (cretz) [inline src/Mutex/WorkflowWithMutex.workflow.cs:9]: Might as well be static

File: src/Mutex/WorkflowWithMutexInput.cs
Comment (cretz) [inline src/Mutex/WorkflowWithMutexInput.cs:3]: Feel free to accept `TimeSpan`s here if you are willing to pass them around for other use cases

**Review by cretz — COMMENTED** [review]
Added some comments, but overall this is the same as I was thinking. I like the idea of wrapping as a utility.

File: src/Mutex/Program.cs
Comment (cretz) [inline src/Mutex/Program.cs:34]: Feel free to have a helper or extension method for registering mutex stuffs instead of a user having to know the things needed

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (devbased) [reply src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:8]: Here a separate object is used, which is returned from WorkflowMutex for the reason that signals need to be sent and they are not strongly typed. Therefore, I thought that it would be better for all internal communication to be centralized in WorkflowMutex. In general, when I was writing this sample, I constantly felt a contradiction, that in some places we have strongly typed calls, while others do not have such capability

File: src/Mutex/Impl/ILockHandle.cs
Comment (devbased) [reply src/Mutex/Impl/ILockHandle.cs:1]: I wanted everything related to the implementation of the mutex to be clearly separated, meaning that potentially it could be a separate project, where only WorkflowMutex.LockAsync (and possibly methods that register workflows/activities) is accessible. On the other hand, since this is just an example, creating multiple projects is overkill

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (devbased) [reply src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:12]: I understood that the constructor is called with an input, and you can populate the Queue there, but RunAsync still needs to have an argument, which is not used if there is a constructor. It looks strange in the end

File: src/Mutex/Impl/ILockHandle.cs
Comment (cretz) [reply src/Mutex/Impl/ILockHandle.cs:1]: Hrmm, yeah, makes sense to separate everything out, just maybe not two levels deep then. I wish I could think of a better word than `Impl` heh.

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [reply src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:8]: Hrmm, yeah it makes sense to me since you're sharing those names. I think I was just hoping to have to jump around to fewer files for this simple sample to see what things do.

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [reply src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs]: Concur, but it should be noted that "signal with start" in Temporal, for special ordering reasons, will call a signal handler _before_ "run" is called. But in your case you only ever send signal-with-start with no initial requests, so you're good I think.

File: src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs
Comment (cretz) [reply src/Mutex/Impl/Internal/MutexWorkflow.workflow.cs:32]: Ah, true, I was thinking of a case where a mutex workflow could lock several independent resources, which doesn't apply here. Can consider this resolved. I think I was confused by the presence of the `ResourceId` in the lock request, which seems like it should be a product of the workflow as a whole.

File: src/Mutex/NotifyLockedInput.cs
Comment (cretz) [inline src/Mutex/NotifyLockedInput.cs:1]: To reduce file count, can put these tiny classes as inner classes of the activity. Not sure I ever would in real code, just trying to trim down the many files people have to jump across to read this sample

**Review by cretz — APPROVED** [review]
Looks good. Marking approved. I think it can be simplified a good bit, though I know that file and package sprawl is kinda the .NET way. Approving for CI run, but of course it will fail due to the relative-path dep, no worry there, we can leave this PR open for a bit.

**Comment by devbased** [comment]
@cretz I reduced the number of files, but I didn't refactor the inputs into nested classes; instead, I placed them outside in the same file where they are being used. Additionally, I provided a bit more detailed description of the project's structure in the README.md

File: src/Mutex/Activities.cs
Comment (cretz) [inline src/Mutex/Activities.cs:3]: Ideally the namespace is at the top, but don't think it matters that much.

File: src/Mutex/Impl/Internal/MutexActivities.cs
Comment (cretz) [reply src/Mutex/Impl/Internal/MutexActivities.cs]: I opened https://github.com/temporalio/sdk-dotnet/issues/134

File: src/Mutex/Impl/LockRequest.cs
Comment (cretz) [inline src/Mutex/Impl/LockRequest.cs:3]: Can we remove the `ResourceId` from this since it's not a property of the request it's a property of the entire workflow? If the resource ID is needed from the workflow, can store on the workflow or derive from workflow ID.

**Review by cretz — APPROVED** [review]
Looks good!

> I didn't refactor the inputs into nested classes; instead, I placed them outside in the same file where they are being used.

I think it's clearer as nested classes (I like the type-per-file rule), but not a big deal I suppose.

Will merge when SDK is updated. Thanks!

File: src/Mutex/Activities.cs
Comment (devbased) [reply src/Mutex/Activities.cs:3]: Currently, `.editorconfig` does not contain `csharp_using_directive_placement`. Should I add `csharp_using_directive_placement = inside_namespace`?

Comment (devbased) [reply src/Mutex/Activities.cs:1]: Cleaned up 'usings'

**Comment by cretz** [comment]
Thanks for the update! Gonna merge #39 first if that's ok (unsure if it'll merge cleanly).

**Comment by cretz** [comment]
Resolved conflict (using GH's built-in conflict editor for the first time, was interesting). Merging once CI passes. Thanks again!

**Pull request merged**
